### DISCUSSION TITLE
Add support for hooks to the SDK or lua api?

### DISCUSSION THREAD

**Discussion by fcbry** [valkey-io/valkey#914] [General]
# Redis Enhancement Proposal: Flexible Event Hooks for Large Object Monitoring

## Background
I recently spent considerable time attempting to implement a feature to log warnings when objects above a certain size (e.g., 1KB) are pushed to Redis. This experience highlighted a potential area where the project could differentiate itself and improve upon the original Redis™.

## Attempted Solutions
1. **Lua Scripting**: Hooks are unsupported (ref: [Redis Programmability](https://redis.io/docs/latest/develop/interact/programmability/))
2. **C Modules**: Using [RedisModulesSDK](https://github.com/RedisLabsModules/RedisModulesSDK) and [Redis Modules API](https://redis.io/docs/latest/develop/reference/modules/), but found lacking support for this use case
3. **Keyspace Notifications**: While available, this solution has limitations:
   - Requires running a separate client to capture pub/sub messages
   - Not suitable for containerized environments (e.g., Kubernetes)
   - Risk of message loss if the client disconnects

## Proposed Enhancement
Extend the RedisModulesSDK to include a flexible, blocking method that allows running a C function when specific operations (e.g., "SET") are executed. This would enable custom monitoring and logging without modifying core Redis source code.

## Benefits
1. Enhanced monitoring capabilities
2. Improved flexibility for developers
3. Maintain compliance and QA standards
4. Better suited for containerized and cloud environments

## Current Workarounds and Their Limitations
- Modifying Redis source code: Not ideal for maintaining QA and compliance standards
- Performance monitoring tools (e.g., perf, eBPF): Often restricted in managed environments due to lack of root access and compliance issues

## Call to Action
Implementing this feature could significantly enhance Redis's capabilities and address a gap in current functionality. It would demonstrate responsiveness to community needs and potentially set Redis apart from other solutions in the market.

We welcome discussion on:
1. The feasibility of implementing this feature
2. Potential design considerations
3. Use cases that would benefit from this enhancement
4. Any concerns or potential drawbacks

Let's collaborate to make Valkey even more powerful and flexible for diverse use cases and environments.

```mermaid
sequenceDiagram
    participant Client
    participant Redis Module
    participant Redis Core

    Client->>Redis Core: SET key value
    Redis Core->>Redis Module: Invoke event hook
    Redis Module->>Redis Module: Check object size
    alt Object size > threshold
        Redis Module->>Redis Module: Log warning
    end
    Redis Module-->>Redis Core: Continue operation
    Redis Core->>Redis Core: Process SET command
    Redis Core-->>Client: OK
```

**Comment by madolson** [comment]
Did you mean to publish this to the Redis repository as opposed to the Valkey one?
